sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4306,8 +4306,14 @@ pi_result piProgramBuild(pi_program Program, pi_uint32 NumDevices,
   // check now for unresolved symbols.
   ze_result_t ZeResult = checkUnresolvedSymbols(ZeModule, &Program->ZeBuildLog);
   if (ZeResult == ZE_RESULT_ERROR_MODULE_LINK_FAILURE) {
+    // remove ZeModule that is associated with the failed program
+    ZE_CALL_NOCHECK(zeModuleDestroy, (ZeModule));
+    ZeModule = nullptr;
```

Comment:
why do you need to zero the local ZeModule before exit?

Reply:
I must have been too tired at late night.
I removed it.
Thanks.

Reply:
We also have a call to zeModuleDestroy in piProgramRelease. Can't this cause a double-free?
Even if piProgramBuild failed we don't exit immediately, exception will be thrown by dpcpp rt which is supposed to be handled afterwards. So, at some point (when sycl::program object is destructed) piProgramRelease may be called which result in double call to zeModuleDestroy.

So maybe we need to assign a nullptr.

Reply:
That is an interesting observation! We still create the PI program, even though return an error. So we expect SYCL RT to call its release, right? Then why can't that release do the destroy, and not here? 

Another observation is that sometimes we return an error *before* we even create a PI object, then how would SYCL RT know if the object was created (and needs a release) in case of such error?

Reply:
No, SYCL RT throws an exception and won't call piProgramRelease() if the build attempt failed.
```
    RT::PiResult Error = Plugin.call_nocheck<PiApiKind::piProgramBuild>(
        Program.get(), /*num devices =*/1, &Device, CompileOptions.c_str(),
        nullptr, nullptr);
    if (Error != PI_SUCCESS)
      throw compile_program_error(getProgramBuildLog(Program.get(), Context),
                                  Error);
    return Program;
```
We would not have the memory leak problem if SYCL RT already calls piProgramRelease for the failed program.

Reply:
> No, SYCL RT throws an exception and won't call piProgramRelease() if the build attempt failed.

The app may catch that exception and then SYCL RT will call the destructor and the piProgramRelease

Reply:
That's not what's happening.
If RT calls piProgramRelease for the failed program, we should not have a memory leak already.

Reply:
I am saying that it is not a true memory leak if the thrown exception is not handled by the app, and execution is terminated.  What if you modify the app to try/catch around the code that yields the error?

Reply:
Yes, correct. So if piProgramCreate is successful then we expect SYCL RT to call piProgramRelease in any case. But zeModuleCreate is called in piProgramBuild. If piProgramBuild is successful then we can't destroy module in piProgramBuild, we need to destroy it in piProgramRelease. If piProgramBuild failed then we can't destroy module in piProgramRelease because we don't know if zeModuleCreate failed or some other error happened - in this case we must destroy module in piProgramBuild. But we need to be sure that zeModuleDestroy is not called again in piProgramRelease if it was alread called in piProgramBuild.
That's why I believe we need zeModuleDestroy in both places and we need to assign nullptr in piProgramBuild to avoid double free. 


I think, if error happened in pi*Create API (i.e. it returned an error code) then SYCL RT have to assume that object was not created and should not call pi*Release. If in pi*Create we created some objects and faced an error afterwards in scope of this pi*Create then we should destroy these objects on our own and return an error code.

Reply:
> If piProgramCreate is successful then we expect SYCL RT to call piProgramRelease in any case

This isn't true. If the exception thrown by SYCL RT is not handled by the app, then no destructor will be called, as no unwinding is going to happen.

Reply:
> If piProgramBuild failed then we can't destroy module in piProgramRelease because we don't know if zeModuleCreate failed or some other error happened - in this case we must destroy module in piProgramBuild. But we need to be sure that zeModuleDestroy is not called again in piProgramRelease if it was alread called in piProgramBuild.

The non-zero ZeModule in PI program should mean the module was created and needs to be destroyed in piProgramRelease. I don't think why we need to destroy it if the later "linking" of that module failed, which we are dealing with here.

Reply:
Ok. it makes sense now. Thanks @smaslov-intel 
Since we call zeModuleDestroy() as follows in ~pi_program() destructor.
```
  if (ZeModule && OwnZeModule) {
    ZE_CALL_NOCHECK(zeModuleDestroy, (ZeModule));
  }
```
Can we set `Program.OwnZeModule = false` when we detect build failed?
This will avoid calling zeModuleDestroy() even if the app catches the exception and calls piProgramRelease.

Reply:
I agree, so we need to destroy only in piProgramRelease. But if module creation somehow failed we need to be sure that it is assigned nullptr (to avoid calling invalid destroy).

Reply:
>  But if module creation somehow failed we need to be sure that it is assigned nullptr (to avoid calling invalid destroy).

Yes, we have to have it null in case no valid destroy on it is possible. Probably here https://github.com/intel/llvm/blob/7f59270e631051b34c0dfa07060a74f76054b6ed/sycl/plugins/level_zero/pi_level_zero.cpp#L4176 (in case zeModuleCreate doesn't put 0 there already in case of an error)

Reply:
@smaslov-intel @againull 
I think we just need to call zeModuleDestroy() and set Program->OwnZeModule = false.
I don't think we can set Program = nullptr because RT still calls Program.get() when it receives PI_FAIL from piProgramBuild.

Reply:
> I don't think we can set Program = nullptr 

Right, we shouldn't. We set Program->ZeModule = nullptr

> and set Program->OwnZeModule = false.

The semantics of the `OwnZeModule` is different, let's not mix it in.

Reply:
Yes, "Program->ZeModule = nullptr" was my original intention of "zeModule = nullptr". :)
Okay, could you confirm if this the right fix?
```
  ze_result_t ZeResult = checkUnresolvedSymbols(ZeModule, &Program->ZeBuildLog);
  if (ZeResult != ZE_RESULT_SUCCESS) {
    // remove ZeModule that is associated with the failed program
    ZE_CALL_NOCHECK(zeModuleDestroy, (ZeModule));
    // Also set Program->ZeModule nullptr to avoid double destroy of zeModule in case
    // where SYCL RT calls piProgramRelease().
    Program->ZeModule = nullptr;
    if (ZeResult == ZE_RESULT_ERROR_MODULE_LINK_FAILURE)
      return PI_BUILD_PROGRAM_FAILURE;
    return mapError(ZeResult);
  }
```

Reply:
No, it is not. You have to run https://github.com/intel/llvm/blob/7f59270e631051b34c0dfa07060a74f76054b6ed/sycl/plugins/level_zero/pi_level_zero.cpp#L4199 to create Program first. Just set ZeModule to null, set error code, and continue (not return).